user/sccsput.man
.\" $Id: sccsput.man,v 6.3 2023/01/20 09:27:57 tom Exp $
.de Es
.ne \\$1
.nr mE \\n(.f
.RS 5n
.sp .7
.nf
.nh
.ta 9n 17n 25n 33n 41n 49n
.ft CW
..
.de Eh
.ft \\n(mE
.fi
.hy \\n(HY
.RE
.sp .7
..
.\" Bulleted paragraph
.de bP
.ie n  .IP \(bu 4
.el    .IP \(bu 2
..
.TH SCCSPUT 1
.SH NAME
.PP
\fBsccsput\fR \- SCCS check-in script
.SH USAGE
.PP
\fBsccsput\fI [options] [file-specifications]\fR
.SH SYNOPSIS
.PP
\fBSccsput\fR is a simple, easy to use interface to \fIsccs\fR
(\fBs\fRource \fBc\fRode \fBc\fRontrol \fBs\fRystem).
For each file specified as input, it checks differences against the
previously archived version and prompts you for change history comments.
.SH DESCRIPTION
.PP
\fBSccsput\fR uses the \fIsccs\fR utilities \fBadmin\fR and \fBdelta\fR
to maintain versions of a given source file in a dependent directory
named "SCCS".
It is more than an integration of the \fBadmin\fR
and \fBdelta\fR utilities, however:
.bP
It checks to ensure that each file is indeed a text
file (so that you do not accidentally archive ".o" files, for
example).
.bP
If you give \fBsccsput\fR a directory name, it will
recur, checking-in files in the directory.
.bP
For each file which has a corresponding "s." file,
\fBsccsput\fR compares the two (using \fBdiff\fR) and pipes the result
through the pager.
.bP
An option is provided so that you may direct \fBsccsput\fR
to perform the differencing without checking the file into sccs.
.bP
The "s." file is post-processed by \fBsccsput\fR
so that the check-in date matches the file's modification date.
.PP
The last point is the fundamental advantage offered by \fBsccsput\fR.
The ordinary \fIsccs\fR methodology uses the current date as the
check-in date.
This works well only for large projects in which a
central project administrator is responsible for controlling the versions
of source files.
It does not work well for small projects,
for which \fIsccs\fR's primary advantage
is its compact storage of multiple versions of a file.
.PP
By using the file's modification date as a reference, you can more
easily back
up to a meaningful version \- by date, rather than version
number.
(By working exclusively in terms of modification date,
you lose the ability to specify sccs release numbers \-
given the complexity
of \fIsccs\fR's interface for release and version numbers, this
is probably not such a great loss).
.PP
\fBSccsput\fR integrates all of the functions used in the \fIsccs\fR
check-in process into one utility program.
.SH OPTIONS
.PP
Some of the options which you may specify to \fBsccsput\fR are
passed through to the underlying utilities.
Others represent extensions:
.TP
-\fBb\fR
is passed to \fBdiff\fR, and directs it to ignore
trailing blanks on a line, and to treat repeated blanks as a single
blank.
.TP
-\fBc\fR
directs \fBsccsput\fR to use \fBcat\fR rather
than the \fBPAGER\fR (usually \fBmore\fR) to display differences.
This is most useful in an Apollo pad, since the \fBmore\fR program
would otherwise switch to VT100 emulator mode.
.TP
\fR-\fBf\fR
forces a check-in, ignoring the output
of the \fBfile\fR utility, which identifies text files.
.TP
-\fBl\fR \fIfile\fR
causes \fBsccsput\fR to generate
a log-file of the files which are processed, and all differences which
are encountered.
The log-file is inherited in recursion to lower
directory levels (i.e., it is written to the same place).
.TP
-\fBn\fR
instructs \fBsccsput\fR to test for differences,
but not to check the files into \fIsccs\fR.
.TP
-\fBs\fR
suppresses some of the messages generated
by the \fIsccs\fR \fBdelta\fR utility describing the number of lines changed,
etc.
.SH OPERATIONS
.PP
The \fBsccsput\fR utility is designed for use in small development
projects.
The methodology for this tool follows:
.bP
Develop source files "normally".
Each file should
contain sccs keywords (see \fIget\ (1)\fR) so that you will be
able to distinguish checked-out files.
The \fIsccs\fR keywords
should appear at the top of your source file, for consistency.
In
C language programs, the convention is to make a string which will
permit the \fBwhat\fR utility to show the versions of the modules
which make up a program:
.Es
#ifndef	lint
static	char	sccs_id[] = "@(#)sccsput.doc	1.1 88/05/05
08:07:16";
#endif
.Eh
.bP
Periodically archive (with \fBsccsput\fR) those
versions of files which you wish to keep
(You should never have programs which have new features
which you wish to keep, while there are defects in other parts of
the program \- that would be an unsound approach to development!).
.bP
When you reach the point of releasing the program, ensure
that all source files have been checked-in.
The directory editor
(\fBded\fR) is useful for reviewing the check-in dates.
.bP
Copy the directory containing your program to the release
directory.
Purge all files, except those which are stored in the \fIsccs\fR subdirectories.
Use \fBsccsget\fR to extract the files
(the unadorned \fBget\fR utility will work, of course,
but it retains the file modification dates).
.bP
Ensure that all files have been checked-in and released.
You may use \fBdiff\fR to compare the directories \- the only
differences should be the substituted \fIsccs\fR keywords.
.bP
Build the released version of your program.
All files
should be present.
No embedded path names should refer to your development
copy.
To ensure good isolation, you may change the permissions on
your development directory temporarily.
.PP
\fBSccsput\fR checks your source file out after the check-in, automatically.
This is done to facilitate development.
A check-in simply adds the
latest changes to a file onto the archive.
.PP
When checking files into \fIsccs\fR, it is a good idea to make
a test run (using the "-\fBn\fR" option) so that you can inspect
the differences.
For example, you may have forgotten to remove (or
bypass) debugging stubs.
Or, you may have been editing a checked-out
file (with the \fIsccs\fR keywords substituted).
\fBSccsput\fR
would archive this anyway.
If you forget, and wish to kill the check-in,
wait until the "comments?" prompt is issued by the \fBdelta\fR
utility.
At this point you may kill \fBsccsput\fR without having
to clean up temporary files.
.PP
If you do not have write-permission on the "SCCS" directory, but
wish to review changes, use the "-\fBn\fR" option.
The intermediate
files are written in the \fB/tmp\fR directory.
.SH ENVIRONMENT
.PP
\fBSccsput\fR is a Bourne shell script.
On Apollo DOMAIN/IX, it
uses System\ 5 features including \fIdirname\ (1)\fR and \fIgetopt\ (1)\fR.
.PP
Environment variables imported by \fBsccsput\fR include:
.PP
\fBSccsput\fR also uses the following environment variables:
.TP
\fBNOTE\fR
Provides a default value for the delta comments.
Normally you should provide case-by-case comments for each file.
This variable is provided so that other programs can invoke \fBsccsput\fR.
If the \fBNOTE\fR variable is defined (i.e., non-null) it is used;
you will not be prompted for comments.
.TP
\fBPAGER\fR
identifies the program to use in displaying differences
between the file which is being checked in, and the previously archived
version.
There may be a lot of differences \- more than can be
shown on one screen.
.TP
\fBSCCS_DIR\fR
specifies the directory into which the \fIsccs\fR "s." files are stored.
If no specified, \fBsccsput\fR
assumes "SCCS".
.SH FILES
.PP
\fBSccsput\fR uses the following files
.TP
\fBsccsput\fR
the Bourne shell script
.TP
\fBputdelta\fR
A utility which invokes \fBadmin\fR
or \fBdelta\fR as required, and modifies the sccs "s."
file after check-in so that the check-in date matches the file's modification
date.
.SH ANTICIPATED CHANGES
.PP
Make \fBsccsput\fR clean up temporary files if it is interrupted.
.PP
Provide a mechanism for inserting dummy version numbers
so that \fBsccsput\fR can bump the release number (for genuine major releases).
Currently,
the SID's are restricted to 1.1, 1.2, 1.3, etc.
.SH SEE ALSO
.PP
putdelta, sccsget, ded, admin\ (1), delta\ (1), diff\ (1), get\ (1),
rmdel\ (1), what\ (1)
.SH AUTHOR
.PP
Thomas Dickey (Software Productivity Consortium).


CHANGES
2023-01-21  Thomas E. Dickey  <dickey@invisible-island.net>

	* bin/makefile.in:
	s/USER/AS_USER/ to avoid environment-variable conflict

	* package/sccs_tools-12.0.spec: amended the version-dependency

2023-01-20  Thomas E. Dickey  <dickey@invisible-island.net>

	* src/getdelta/test/run_test.sh, src/sccs2rcs/test/run_test.sh,
	  src/sccs2rcs/test/do_test.sh:
	shellcheck warnings

	* src/putdelta/test/run_test.sh: fix shellcheck warnings

	* src/putdelta/test/run_test.sh:
	discard stderr (some variants complain profusely about missing keywords)

	* src/getdelta/test/run_test.sh:
	modify to ensure the files have SCCS keywords, since some variants do not handle this gracefully

	* user/sccs2rcs.man: check-manpage

	* user/fixsccs.man: RCS_BASE

	* user/sccsget.man, user/sccsput.man, user/getdelta.man,
	  user/putdelta.man:
	check-manpage

	* user/makefile.in: add manpage for fixsccs

	* package/debian/changelog, package/sccs_tools-12.0.spec: bump

	* package/debian/control: update standards

	* package/debian/compat: bump

2023-01-19  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/sccs_tools-12.0.spec:
	use td_lib package, rather than side-by-side configuration

	* configure: regen

	* aclocal.m4: provide for finding td_lib using pkg-config

	* package/debian/copyright, COPYING: bump

2022-12-31  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/debian/changelog, package/sccs_tools-12.0.spec: bump

	* configure: regen

	* aclocal.m4: resync with my-autoconf

2022-10-11  Thomas E. Dickey  <dickey@invisible-island.net>

	* user/makefile.in: use install-sh if needed

	* src/putdelta/test/run_test.sh: s/fgrep/grep/

	* package/debian/copyright: bump

	* configure: regen

	* aclocal.m4: resync with my-autoconf

	* COPYING, package/debian/changelog, package/sccs_tools-12.0.spec: bump

2022-08-01  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess, config.sub: 2022-08-01

2021-03-28  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/debian/changelog, package/sccs_tools-12.0.spec: bump

	* configure: regen

	* aclocal.m4: resync with my-autoconf

2021-01-10  Thomas E. Dickey  <dickey@invisible-island.net>

	* configure: regen

	* aclocal.m4: quoting fixes for shellcheck

	* src/getdelta/src/getdelta.c, src/getdelta/src/sccsget.c,
	  src/sccs2rcs/src/sccs2rcs.c:
	gcc-warnings

	* configure: regen

	* aclocal.m4: resync with my-autoconf

	* COPYING, package/debian/changelog, package/debian/copyright,
	  package/sccs_tools-12.0.spec:
	bump

2020-12-08  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/debian/changelog, package/sccs_tools-12.0.spec: bump

2020-12-07  Thomas E. Dickey  <dickey@invisible-island.net>

	* src/getdelta/test/makefile.in, src/putdelta/test/makefile.in,
	  src/sccs2rcs/test/makefile.in, src/fixsccs/test/makefile.in,
	  src/getdelta/src/makefile.in, src/putdelta/src/makefile.in,
	  src/sccs2rcs/src/makefile.in, src/fixsccs/src/makefile.in:
	use VPATH

2020-12-06  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/debian/copyright, COPYING, package/debian/changelog,
	  package/sccs_tools-12.0.spec:
	bump

2020-11-30  Thomas E. Dickey  <dickey@invisible-island.net>

	* install-sh: update to 2020-11-14

2020-11-26  Thomas E. Dickey  <dickey@invisible-island.net>

	* configure: regen

	* aclocal.m4: resync with my-autoconf

	* configure.in: without-X

2020-08-16  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess, config.sub: 2020-08-17

2019-12-21  Thomas E. Dickey  <dickey@invisible-island.net>

	* src/putdelta/src/sccsput.c: suppress pager if not a tty

	* package/debian/changelog, package/sccs_tools-12.0.spec: bump

	* src/putdelta/src/putdelta.c: add an fflush

2019-12-06  Thomas E. Dickey  <dickey@invisible-island.net>

	* user/putdelta.man: mention the other options that putdelta handles

	* src/sccs2rcs/src/sccs2rcs.c: stricter gcc warnings

	* src/putdelta/src/putdelta.c, src/putdelta/src/sccsput.c:
	use DYN-argv lists to replace catarg()

	* package/debian/compat: bump

	* src/getdelta/src/getdelta.c: use DYN-argv lists to replace catarg()

	* user/sccsget.man, user/sccsput.man, user/putdelta.man,
	  user/getdelta.man:
	check-manpage fixes (rename Ex to Eh for consistency, other errata)

	* src/putdelta/test/run_test.sh: use sccs wrapper

	* src/getdelta/src/sccsget.c: use DYN-argv lists to replace catarg()

	* src/getdelta/test/run_test.sh: use sccs wrapper

	* configure: regen

	* aclocal.m4: resync with my-autoconf

	* package/debian/copyright, COPYING, package/debian/changelog,
	  package/sccs_tools-12.0.spec:
	bump

2019-09-11  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess, config.sub: 2019-09-10

2018-03-24  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/debian/changelog: bump

	* package/sccs_tools-12.0.spec:
	bump, disable debug-packages, update ftp url

	* package/debian/watch: update ftp url

	* package/debian/rules: use Debian build-flags

2018-01-07  Thomas E. Dickey  <dickey@invisible-island.net>

	* package/debian/copyright, COPYING, package/debian/changelog,
	  package/sccs_tools-12.0.spec:
	bump

	* configure: regen

	* aclocal.m4, configure.in: resync with my-autoconf

2017-11-25  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess: 2017-11-07

	* config.sub: 2017-11-23

2015-07-04  Thomas E. Dickey  <dickey@invisible-island.net>

	* user/makefile.in: improve clean-rule

	* package/sccs_tools-12.0.spec, package/debian/copyright,
	  package/debian/changelog, COPYING:
	bump

	* configure: regen

	* aclocal.m4: resync with my-autoconf

2015-05-02  Thomas E. Dickey  <dickey@invisible-island.net>

	* config.guess: 2015-03-04

	* config.sub: 2015-03-08

2014-12-14  Thomas E. Dickey  <dickey@invisible-island.net>

	* user/sccsput.man, user/putdelta.man, user/getdelta.man,
	  user/sccsget.man:
	adapt/reuse macros from cm_tools to improve groff's translation to html

2013-12-07  Thomas E. Dickey  <dickey@invisible-island.net>


src/putdelta/src/sccsput.c
/*
 * Title:	sccsput.c (sccs put-tree)
 * Author:	T.E.Dickey
 * Created:	08 May 1990 (from sccsput.sh and rcsput.c)
 * Modified:
 *		21 Dec 2019, suppress pager if not a tty.
 *		06 Dec 2019, use DYN-strings to replace catarg.
 *		14 Oct 1995, allow archive leafname to be limited to 14-chars
 *		27 May 1994, added "-e" and "-C" options.  Made verbosity
 *			more consistent with CM_TOOLS.
 *		23 Sep 1993, gcc warnings
 *		24 Oct 1991, converted to ANSI
 *		13 Sep 1991, use common 'filesize()'
 *		24 Jul 1991, corrected size of 'comment[]'
 *		22 Jul 1991, cleanup use of 'catarg()'
 *		19 Jul 1991, accept "-r" option (for diff and putdelta)
 *		18 Jul 1991, renamed "-f" option to "-F", added new "-f" to
 *			pass-down to 'putdelta'.  Also, pass-thru "-k" to
 *			'putdelta'.
 *		25 Jun 1991, added "-D", "-T" options.  Automatically unlink
 *			log-file if it is empty (may retain prior file if one
 *			exists).
 *		20 Jun 1991, use 'shoarg()'.
 *		07 Jun 1991, use 'getopt()'; cleanup option processing,
 *			including those that we pass-thru to 'putdelta'
 *			corrected pathname of "-l" value
 *		
 * Function:	Use 'putdelta' to archive one or more files from the
 *		SCCS-directory which is located in the current working
 *		directory, and then, to set the delta date of the  checked-in
 *		files according to the last modification date (rather than the
 *		current date, as SCCS assumes).
 *
 * Options:	see 'usage()'
 */

#define	STR_PTYPES
#include	<ptypes.h>
#include	<dyn_str.h>
#include	<rcsdefs.h>
#include	<sccsdefs.h>

MODULE_ID("$Id: sccsput.c,v 6.15 2019/12/21 23:59:09 tom Exp $")

#define	DEBUG		if (debug) PRINTF
#define	VERBOSE		if (!quiet) PRINTF

static ARGV *diff_opts;
static const char *verb = "putdelta";
static char comment[BUFSIZ];
static FILE *log_fp;
static int a_opt;		/* all-directory scan */
static int no_op;		/* no-op mode */
static const char *pager;	/* nonzero if we don't cat diffs */
static int Force, force;
static int e_opt;
static const char *k_opt = "";
static const char *r_opt = "";
static char *LogName = NULL;
static int debug;
static int quiet;
static int found_diffs;		/* true iff we keep logfile */

static void
ExitProgram(int code)
{
    if (log_fp != 0) {
	FCLOSE(log_fp);
	if (!found_diffs)
	    (void) unlink(LogName);
    }
    exit(code);
}

static void
cat2fp(FILE *fp, const char *name)
{
    FILE *ifp;
    char t[BUFSIZ];
    size_t n;

    if ((ifp = fopen(name, "r")) != NULL) {
	while ((n = fread(t, sizeof(char), sizeof(t), ifp)) > 0)
	    if (fwrite(t, sizeof(char), n, fp) != n)
		  break;
	FCLOSE(ifp);
    }
}

static int
pipe2file(DYN * cmd, char *name)
{
    FILE *ifp, *ofp;
    char buffer[BUFSIZ];
    int empty = TRUE;
    size_t n;

    if (!tmpnam(name))
	failed("tmpnam");
    if (!(ofp = fopen(name, "w")))
	failed("tmpnam-open");
    if (debug) {
	FORMAT(buffer, "%s > ", dyn_string(cmd));
	shoarg(stdout, buffer, name);
    }
    if (!(ifp = popen(dyn_string(cmd), "r")))
	failed("popen");
    /* copy the result to a file so we can send it two places */
    while ((n = fread(buffer, sizeof(char), sizeof(buffer), ifp)) > 0) {
	if (fwrite(buffer, sizeof(char), n, ofp) != n)
	      break;
	empty = FALSE;
    }
    (void) pclose(ifp);
    FCLOSE(ofp);
    return (!empty);
}

static int
different(const char *working, const char *archive)
{
    static char format[] = "------- %s -------\n";
    ARGV *args;
    DYN *str;
    char in_diff[MAXPATHLEN];
    char out_diff[MAXPATHLEN];
    int changed;

    args = sccs_argv("get");
    argv_append(&args, "-p");
    if (!debug)
	argv_append(&args, "-s");
    if (*k_opt)
	argv_append(&args, k_opt);
    if (*r_opt)
	argv_append(&args, r_opt);
    argv_append(&args, archive);
    str = argv_flatten(args);
    (void) pipe2file(str, in_diff);
    argv_free(&args);
    dyn_free(str);

    args = argv_init1("diff");
    argv_merge(&args, diff_opts);
    argv_append(&args, in_diff);
    argv_append(&args, working);
    str = argv_flatten(args);
    changed = pipe2file(str, out_diff);
    argv_free(&args);
    dyn_free(str);

    if (changed) {
	if (!quiet) {
	    if (pager == 0 || !isatty(fileno(stdout))) {
		PRINTF(format, working);
		cat2fp(stdout, out_diff);
	    } else {
		if (execute(pager, out_diff) < 0)
		    failed(pager);
	    }
	}
	if (log_fp != 0) {
	    VERBOSE("... appending to logfile");
	    FPRINTF(log_fp, format, working);
	    cat2fp(log_fp, out_diff);
	    if (!found_diffs && filesize(out_diff) > 0)
		found_diffs = TRUE;
	}
    } else {
	VERBOSE("*** no differences found ***\n");
    }

    (void) unlink(in_diff);
    (void) unlink(out_diff);
    if (e_opt && !changed)
	ExitProgram(FAIL);
    return (changed);
}

static int
ok_file(const char *name)
{
    if (!Force && !istextfile(name)) {
	PRINTF("*** \"%s\" does not seem to be a text file\n", name);
	return FALSE;
    }
    return TRUE;
}

static void
SccsPut(const char *path, const char *name)
{
    ARGV *args;
    const char *working = sccs2name(name, FALSE);
    const char *archive = name2sccs(name, FALSE);
    int first;
#if defined(S_FILES_14)
    char temp[MAXPATHLEN];
#endif

    /*
     * Start by assuming that if we find a non-empty file with the correct
     * name for an archive, that this isn't the first delta.
     */
    if (filesize(archive) >= 0) {
	first = FALSE;
    }
#if defined(S_FILES_14)
    else if (fleaf14(strcpy(temp, archive))
	     && filesize(temp) >= 0) {
	archive = temp;
	first = FALSE;
    }
#endif
    else {
	first = TRUE;
    }

    /*
     * For the initial insertion, verify that the file looks like text.
     * Otherwise we'll catch garbage files when differencing them.
     *
     * If we've found something that looks like an archive, check if it's
     * got at least one version.
     */
    if (first) {
	if (!ok_file(working))
	    return;
    } else {
	time_t date;
	const char *vers, *locker;
	sccslast(path, working, &vers, &date, &locker);
	if (*vers == '?') {
	    first = TRUE;	/* no revisions present */
	    if (!ok_file(working))
		return;
	} else {
	    if (!Force && !different(working, archive))
		return;
	}
    }

    args = argv_init1(verb);
    if (!debug)
	argv_append(&args, "-s");
    if (force)
	argv_append(&args, "-f");
    if (*k_opt)
	argv_append(&args, k_opt);
    if (*r_opt)
	argv_append(&args, r_opt);
    if (*comment)
	argv_append(&args, comment);
    argv_append(&args, working);

    if (!no_op) {
	VERBOSE("*** %s \"%s\"\n",
		first ? "Initial SCCS insertion of"
		: "Applying SCCS delta to",
		name);
	if (debug)
	    show_argv(stdout, argv_values(args));
	if (executev(argv_values(args)) < 0)
	    failed(working);
    } else {
	VERBOSE("--- %s \"%s\"\n",
		first ? "This would be initial for"
		: "Delta would be applied to",
		name);
    }
    argv_free(&args);
}

/*ARGSUSED*/
static int
WALK_FUNC(scan_tree)
{
    char tmp[MAXPATHLEN];
    char *s = pathcat(tmp, path, name);

    (void) level;

    if (sp == 0 || readable < 0) {
	readable = -1;
	perror(name);
	if (!Force)
	    exit(FAIL);
    } else if (isDIR(sp->st_mode)) {
	abspath(s);		/* get rid of "." and ".." names */
	if (!a_opt && *pathleaf(s) == '.')
	    readable = -1;
	else if (sameleaf(s, sccs_dir((char *) 0, (char *) 0))
		 || sameleaf(s, rcs_dir(NULL, NULL)))
	    readable = -1;
	else if (debug)
	    track_wd(path);
    } else if (isFILE(sp->st_mode)) {
	if (debug)
	    track_wd(path);
	SccsPut(path, name);
    } else
	readable = -1;

    return (readable);
}

static void
do_arg(const char *name)
{
    (void) walktree((char *) 0, name, scan_tree, "r", 0);
}

static void
usage(int option)
{
    static const char *tbl[] =
    {
	"Usage: sccsput [options] files_or_directories"
	,""
	,"Options"
	,"  -a       process all directories, including those beginning with \".\""
	,"  -b       (passed to \"diff\")"
	,"  -C       (passed to \"diff\")"
	,"  -c       send differences to terminal without $PAGER filtering"
	,"  -e       error if no differences"
	,"  -f       force (passed to putdelta)"
	,"  -h       (passed to \"diff\")"
	,"  -k       suppress keyword expansion from archive before differences"
	,"           (also passed to putdelta; implies file has unexpanded keywords)"
	,"  -l file  write all differences to log-file"
	,"  -n       compute differences only, don't try to archive"
	,"  -r SID   specify SCCS-sid (version)"
	,"  -s       silent (passed to putdelta)"
	,"  -y text  comment (passed to putdelta)"
	,"  -D opts  special options to pass to \"diff\""
	,"  -F       force (initial) insertion even if file appears to be non-text"
	,"  -T tool  specify a checkin-tool other than \"putdelta\""
	,""
    };
    unsigned j;
    for (j = 0; j < sizeof(tbl) / sizeof(tbl[0]); j++)
	FPRINTF(stderr, "%s\n", tbl[j]);
    if (option == '?') {
	ARGV *args = argv_init1("putdelta");
	argv_append(&args, "-?");
	executev(argv_values(args));
    }
    exit(FAIL);
}

/*ARGSUSED*/
_MAIN
{
    char deferred[BUFSIZ];
#define	DEFERRED	strcat(strcpy(deferred, "-"), optarg)

    int j;

    revert("not a setuid-program");	/* ensure this is not misused */
    pager = dftenv("more -l", "PAGER");
    debug = sccs_debug();
    while ((j = getopt(argc, argv, "abCcefhkl:nr:sy:D:FT:")) != EOF) {
	switch (j) {
	case 'a':
	    a_opt = TRUE;
	    break;
	case 'b':
	    argv_append(&diff_opts, "-b");
	    break;
	case 'C':
	    argv_append(&diff_opts, "-c");
	    break;
	case 'c':
	    pager = 0;
	    break;
	case 'e':
	    e_opt = TRUE;
	    break;
	case 'f':
	    force = TRUE;
	    break;
	case 'h':
	    argv_append(&diff_opts, "-h");
	    break;
	case 'k':
	    k_opt = "-k";
	    break;
	case 'l':
	    found_diffs = filesize(LogName = optarg) > 0;
	    if (!(log_fp = fopen(optarg, "a+")))
		usage(0);
	    break;
	case 'n':
	    no_op = TRUE;
	    break;
	case 'r':
	    FORMAT(deferred, "-r%s", optarg);
	    r_opt = stralloc(deferred);
	    break;
	case 's':
	    quiet = TRUE;
	    break;
	case 'y':
	    FORMAT(comment, "-y%.*s",
		   (int) (sizeof(comment) - 3), optarg);
	    break;
	case 'D':
	    argv_append(&diff_opts, DEFERRED);
	    break;
	case 'F':
	    Force = TRUE;
	    break;
	case 'T':
	    verb = optarg;
	    break;
	default:
	    usage(j);
	}
    }

    if (debug)
	track_wd((char *) 0);
    if (optind < argc) {
	while (optind < argc)
	    do_arg(argv[optind++]);
    } else {
	do_arg(".");
    }

    ExitProgram(SUCCESS);
    /*NOTREACHED */
}
